docs/madr/decisions/038-transition-to-new-policies.md

```diff
@@ -0,0 +1,473 @@
+# Transition to new policies
+
+- Status: accepted
+
+## Context and Problem Statement
+
+The current state of the policies as of Kuma 2.6.x:
+* new policies are in GA, they're a recommended way to configure new deployments in Kuma
+* old policies are not deprecated, they're still in use by majority of Kuma users
+
+The goal is to provide a way to transition from old policies to new policies. Having a smooth transition is going to enable 
+us to deprecate old policies in the future.
+
+Prerequisites for the transition:
+* both Global and Zone CPs are running the latest Kuma version
+
+## Decision Drivers
+
+- Predictability: when replacement policies are created (manually or automatically), user should be able to verify them 
+without applying for real
+- Short feedback loop: we'd like to deliver small portions of new policies more often, rather than big portions but rarely
+- Quick rollback: user should be able to rollback to old policies
+
+## Considered Options
+
+- Introduce a shadow mode for policies on Zone CP ✅ / ❌
+- Introduce a shadow mode for policies on Global CP ✅ / ❌
+  - Show potential changes for a specific DPP ✅ / ❌
+  - Show potential changes for all affected DPPs ✅ / ❌
+
+## Shadow mode for policies
+
+Shadow mode is a way to test new policies without affecting the real traffic. When the policy is created with `kuma.io/effect: shadow` label
+
+```yaml
+type: MeshTimeout
+name: timeout-global
+mesh: default
+labels:
+  kuma.io/effect: shadow
+spec:
+  targetRef:
+    kind: Mesh
+  to: [...]
+```
+
+it doesn't apply on the real proxies. Instead, by using a special endpoint we can get a list of jsonPatches that would be applied to the Envoy configuration
+
+```shell
+$ curl :5681/meshes/defaults/dataplane/dpp-1/xds?shadow=true # show potential changes for a dpp-1 in the "default" mesh
+```
+
+Implementation wise, shadow mode requires generating 2 snapshots: one with and without the new policy. 
```

Comment:
How would it behave if we have one MeshTimeout policy (global) with shadow and another MeshTimeout without shadow. After merge is it going to be shadow or not?

Reply:
I want proper testing here. Because I'm not 100% sure the composition works this way:

However is the output a+b(shadow)+c(shadow) useful to figure what a+b+c(shadow) does?
May it's not an actual problem.

What I mean is that I'd like to see more advanced experiments than just using 2 policies.

Reply:
> However is the output a+b(shadow)+c(shadow) useful to figure what a+b+c(shadow) does?
> May it's not an actual problem.

There is no composition for s&d policies right? I might be getting this wrong but what I would image happening is:
1. You have s&d policies (in normal mode because there is no other mode)
2. You add a couple of targetRef policies in shadow mode (might be even doing doing some composition)
3. We generate normal snapshot for s&d policies and another snapshot that has both s&d policies and targetRef policies and we call this snapshot "shadow snapshot"
4. We show the diff to the user, if they like the diff they can remove "shadow" from all policies that have that 
5. We repeat 2-4 until everything is migrated
6. Most important thing and a bit of leap of faith, users remove the old s&d policies. This **heavily** relies on our assumption that targetRef policies take precedence over s&d policies. If this is broken anywhere then this step might mess some things up.